specs/data-grid/data-grid.spec.md

```diff
@@ -0,0 +1,302 @@
+
+```html
+    <template>
+        <button @click="${x => myCustomHandler()}">
+            ${x =>
+                x.rowData === null ||
+                x.columnData === null ||
+                x.columnData.columnDataKey === null
+                    ? null
+                    : x.rowData[x.columnData.columnDataKey]}
+        </button>
+    </template>
+```
+
+
+
+**Data grid**
+- `fast-data-grid`
+
+*Attributes:*
+- `generateHeader`  
+Boolean.  Automatically generate a header element based on provided columns. Default is true.  Authors who wish to not have a header row or wish to generate their own can set this to false.
+
+*properties:*
+- `rowsData`  
+An array of objects that contain the data to be displayed.  Each object corresponds to one row.
+
+- `columnsData`  
+An array of `DataGridColumn` objects that define what columns will be displayed in the grid.  The order of the columns determines their order in the grid.
+
+- `rowItemTemplate`  
+Custom [template](https://fast.design/docs/fast-element/declaring-templates) to use when generating rows by iterating over data.  The default template uses `fast-data-grid-row`, this is where authors change that.
+
+*Slots:*
+- `default`  
+Custom generated rows can be placed here
+
+- `header`
+Items added here are added to the header region of the component.
+
+*Functions:*
+- `generateColumns(object): DataGridColumn`   
+Static function that creates a basic set of columns from an object representing a row.  
+
+*Events*
+- none
+
+*parts:*
+- `headerSlot`
+
+- `rowsSlot`
+
+
+**Data grid header**
+- `fast-data-grid-header`
+
+*Attributes:*
+- none
+
+*properties:*
+- `columnsData`  
+An array of `DataGridColumn` objects that define what columns will be displayed in the grid.  The order of the columns determines their order in the grid.
+
+- `headerCellItemTemplate`  
+Custom [template](https://fast.design/docs/fast-element/declaring-templates) to use when generating header cells by iterating over data.  The default template uses `fast-data-grid-header-cell`, this is where authors change that.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `headerCellsSlot`
+
+
+**Data grid header cell**
+- `fast-data-grid-header-cell`
+
+*Attributes:*
+- extends HTML Element attributes
+
+*properties:*
+- `columnData`  
+The `DataGridColumn` this column header represents.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `headerCellsSlot`
+
+
+**Data grid row**
+- `fast-data-grid-row`
+
+*Attributes:*
+- `gridTemplateColumns`  
+String that gets applied to the the css gridTemplateColumns attribute for the row
+
+*properties:*
+- `rowData`  
+The objects that contains the data to be displayed in this row.
+
+- `columnsData`  
+An array of `DataGridColumn` objects that define what columns will be displayed in the grid.  The order of the columns determines their order in the grid.
+
+- `cellItemTemplate`  
```

Comment:
I'm a bit confused about how all these elements come together. This `fast-data-grid-row` has a default slot to hold items  (items exist as child elements of a row) but it also expects column data and a template to programmatically render the items? Is the intent that either mechanism could be used to render items?

Reply:
Yes, either can work.  An author could choose to just provide data to the top level grid and have it generate based on that, or could create custom rows and cells using the slots.  I expect the data driven scenario to be the primary one.